plugin/src/main/java/org/opensearch/ml/model/MLModelAutoReDeployer.java

```diff
@@ -0,0 +1,471 @@
+            return;
+        }
+
+        String localNodeId = nodeEnvironment.nodeId();
+        // if the local node are both ml and data node,it should throw MLException
+        if (isNode(ML_ROLE_NAME) && isNode(DATA_ROLE_NAME)) {
+            throw new MLException("the node " + localNodeId + " shouldn't be both ml and data node");
+        }
+
+        // auto redeploy all models of this local node
+        threadPool.executor(DEPLOY_THREAD_POOL).execute(() -> {
+            long sleepTimeUnit = 1000L;
+            int numRetries = 4;
+            int numTries = 0;
+
+            while (numTries < numRetries) {
+                ++numTries;
+                log.info("trying to auto-redeploy model. try {} of {}", numTries, numRetries);
+
+                try {
+                    TimeUnit.MILLISECONDS.sleep(sleepTimeUnit * numTries);
+                    autoReDeployModel(localNodeId);
+                    break;
+                } catch (InterruptedException | RuntimeException e) {
+                    log.error("exception while auto-redeploying model.", e);
+                }
+            }
+        });
+    }
+
+    /**
+     * auto re-deploy model under the local node id<br>
+     * @param localNodeId local node id
+     * @throws MLException exception
+     */
+    @VisibleForTesting
+    void autoReDeployModel(String localNodeId) throws MLException {
+        // wait the index ".plugins-ml-model" finish initialization and get all the models to be deployed
+        StepListener<SearchResponse> queryModelStep = new StepListener<>();
+
+        queryModels(ActionListener.wrap(queryModelStep::onResponse, queryModelStep::onFailure));
+
+        List<String> modelIds = new ArrayList<>();
+        queryModelStep.whenComplete(searchResponse -> {
+            SearchHit[] hits = searchResponse.getHits().getHits();
+            if (CollectionUtils.isEmpty(hits)) {
+                // no model need to be deployed
+                return;
+            }
+
+            for (SearchHit hit : hits) {
+                try (XContentParser parser = createXContentParserFromRegistry(xContentRegistry, hit.getSourceRef())) {
+                    ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.nextToken(), parser);
+                    MLModel mlModel = MLModel.parse(parser);
+
+                    if (!mlModel.isDeployToAllNodes()) {
+                        String[] planningWorkerNodes = mlModel.getPlanningWorkerNodes();
+                        if (!ArrayUtils.contains(planningWorkerNodes, localNodeId)) {
+                            continue;
+                        }
+                    }
+
+                    modelIds.add(mlModel.getModelId());
+                } catch (Exception e) {
+                    throw new MLException("failed to parse ml model", e);
+                }
+            }
+
+            modelIds.forEach(modelId -> log.info("the model id {} need to deploy or undeploy", modelId));
```

Comment:
This log looks confusing, let's remove this log line or make it clear, whether the model need to deploy or undeploy. I would suggest remove this log and add logs before line 204, 210.

Reply:
has moved